contracts/libraries/NFTDescriptor.sol

```diff
@@ -0,0 +1,305 @@
+        // start index (inclusive) for trailing/leading 0's for very small/large numbers
+        uint8 zerosStartIndex;
+        // end index for (exclusive) trailing/leading 0's for very small/large numbers
+        uint8 zerosEndIndex;
+        // true if decimal number is less than one
+        bool isLessThanOne;
+        // true if string should include "%"
+        bool isPercent;
+    }
+
+    function generateDecimalString(DecimalStringParams memory params) private pure returns (string memory) {
+        bytes memory buffer = new bytes(params.bufferLength);
+        if (params.isPercent) {
+            buffer[buffer.length - 1] = '%';
+        }
+        if (params.isLessThanOne) {
+            buffer[0] = '0';
+            buffer[1] = '.';
+        }
+
+        // add leading/trailing 0's
+        for (uint256 zerosCursor = params.zerosStartIndex; zerosCursor < params.zerosEndIndex; zerosCursor++) {
+            buffer[zerosCursor] = bytes1(uint8(48));
+        }
+        // add sigfigs
+        while (params.sigfigs > 0) {
+            if (params.decimalIndex > 0 && params.sigfigIndex == params.decimalIndex) {
+                buffer[params.sigfigIndex--] = '.';
+            }
+            buffer[params.sigfigIndex--] = bytes1(uint8(uint256(48).add(params.sigfigs % 10)));
+            params.sigfigs /= 10;
+        }
+        return string(buffer);
+    }
+
+    function addressToString(address addr) internal pure returns (string memory) {
+        uint256 _uint = (uint256(addr));
+        return _uint.toHexString(20);
+    }
+
+    function tickToDecimalString(
+        int24 tick,
+        int24 tickSpacing,
+        uint8 token0Decimals,
+        uint8 token1Decimals
+    ) internal pure returns (string memory) {
+        if (tick == (TickMath.MIN_TICK / tickSpacing) * tickSpacing) {
+            return 'MIN';
+        } else if (tick == (TickMath.MAX_TICK / tickSpacing) * tickSpacing) {
+            return 'MAX';
+        } else {
+            return fixedPointToDecimalString(TickMath.getSqrtRatioAtTick(tick), token0Decimals, token1Decimals);
+        }
+    }
+
+    function sigfigsRounded(uint256 value, uint8 digits) private pure returns (uint256, bool) {
+        bool extraDigit;
+        if (digits > 5) {
+            value = value.div((10**(digits - 5)));
+        }
+        bool roundUp = value % 10 > 4;
+        value = value.div(10);
+        if (roundUp) {
+            value = value + 1;
+        }
+        // 99999 -> 100000 gives an extra sigfig
+        if (value == 100000) {
+            value /= 10;
+            extraDigit = true;
+        }
+        return (value, extraDigit);
+    }
+
+    function adjustForDecimalPrecision(
+        uint160 sqrtRatioX96,
+        uint8 token0Decimals,
+        uint8 token1Decimals
+    ) private pure returns (uint160) {
+        uint256 sqrt10X128 = 1076067327063303206878105757264492625226;
+        uint256 difference = decimalDifference(token0Decimals, token1Decimals);
```

Comment:
cleaned up a bit, but the reason I did this is so I can confirm both these conditions before proceeding: `difference > 0 && difference <= 18` or else I'd have to check twice

Reply:
as in
```
if (token0Decimals < token1Decimals) {
  uint8 difference = token1Decimals - token0Decimals;
  // ...
} else if (token1Decimals < token0Decimals) {
  uint8 difference = token0Decimals - token1Decimals;
  // ...
} else {
  return sqrtRatioX96;
}
```

IMO this is still better in terms of readability

Reply:
even with all this?
```
if (token0Decimals < token1Decimals) {
  uint8 difference = token1Decimals - token0Decimals;
  if (difference > 0 && difference < 18{
    // 
  }
} else if (token1Decimals < token0Decimals) {
  uint8 difference = token0Decimals - token1Decimals;
    if (difference > 0 && difference < 18{
    // 
  }
} else {
  return sqrtRatioX96;
}
```

Reply:
you can just use `SafeMath` to do the scaling, instead of limiting the decimal difference, since there will be some prices that can be supported with greater than 18 decimal difference